### PR TITLE
htlcswitch: sync local payment hand-off to link

### LINKED ISSUE #4181: routing: pathfind with up-to-date bandwidth hints
In the current implementation, a local payment is handed off from switch to link in an asynchronous matter via a so-called mailbox. This means that the call to the switch returns before the outgoing htlc is locked into the channel. If the htlc isn't locked in, the link will report that bandwidth as still being available.

This can cause a problem when multiple pathfinding operations are launched in quick succession. The next pathfinding round queries the link for bandwidth and gets returned a value that does not yet include a previously launched htlc. It looks as if more bandwidth is available than there really is. If the found route is then used for the htlc, it may fail on our own local channel balance insufficiency. No penalties (to ourselves) will be applied though and the pathfinding is retried. By that time, the bandwidth has likely been updated and a valid route is found. It is a temporary problem that causes our operation to be less efficient.

A typical situation where this may happen is when multi-part payments are sent. For multi-part payments, several pathfinding operations may be started in a short time frame.

Before creating a PR, it is worth to investigate possible solutions in this issue. Options that have come up so far:
* Add a mechanism (channel?) that signals back to the router that an htlc has been locked into the channel and that the reported bandwidth is now up to date. It doesn't make the whole system simpler unfortunately.
* Hand-off local payments to the link synchronously.
* ... ?

### PR DESCRIPTION
**Pull request by joostjager**
Experiment to see how synchronous hand-off to the link could work.

Tried it out on the mpp send payment itest. To make sure that the async hand-off would lead to an extra payment attempt, I added a 100 ms sleep in the link before adding the htlc to the channel.

This is the original timing diagram, which shows the temp. chan failure originating from ourselves (@ 0)

image

With sync hand-off, it looks like this:

image

Fixes #4181

### CODE REVIEW

**Comment by joostjager** [comment]
Current problem is that the link balance isn't yet updated after `channel.AddHtlc` has been called.

**Comment by joostjager** [comment]
>Current problem is that the link balance isn't yet updated after channel.AddHtlc has been called.

Faulty analysis. Nothing wrong with calculating the balance, I just printed the wrong number.

**Comment by joostjager** [comment]
No attention paid to unit tests and itests yet. Looking for a concept ack.

**Comment by joostjager** [comment]
Wondering if there is a good reason to pass local sends through the event loop?

File: htlcswitch/link.go [hunk 1]
```diff
@@ -346,6 +351,8 @@ type channelLink struct {
 	// by the HTLC switch.
 	downstream chan *htlcPacket
 
+	downstreamIn chan *downstreamMsg
```
Comment (cfromknecht) [inline htlcswitch/link.go:354]: came could be more descriptive, e.g `localDownstream`

File: htlcswitch/link.go [hunk 2]
```diff
@@ -395,11 +402,12 @@ func NewChannelLink(cfg ChannelLinkConfig,
...
-		hodlMap:     make(map[channeldb.CircuitKey]hodlHtlc),
-		hodlQueue:   queue.NewConcurrentQueue(10),
-		log:         build.NewPrefixLog(logPrefix, log),
-		quit:        make(chan struct{}),
+		htlcUpdates:  make(chan *contractcourt.ContractUpdate),
+		hodlMap:      make(map[channeldb.CircuitKey]hodlHtlc),
+		hodlQueue:    queue.NewConcurrentQueue(10),
+		log:          build.NewPrefixLog(logPrefix, log),
+		quit:         make(chan struct{}),
+		downstreamIn: make(chan *downstreamMsg),
```
Comment (cfromknecht) [inline htlcswitch/link.go:410]: group with other downstream chnanel

File: htlcswitch/switch.go [hunk 3]
```diff
@@ -451,7 +451,26 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
+		return err
+	}
+
+	// Drop duplicate packet if it has already been seen.
+	switch {
+	case len(actions.Drops) == 1:
+		return ErrDuplicateAdd
+
+	case len(actions.Fails) == 1:
+		return err
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply htlcswitch/switch.go:467]: Yup, maybe return error to caller as o/w the `PaymentAttemptDispatcher` will log a successful send to the switch

**Comment by Crypt-iQ** [comment]
Solid set of changes, the way you changed the switch was straightforward and actually reduced complexity. Before I read through the changes, I was worried there would be more complexity. LGTM upon travis pass

**Comment by joostjager** [comment]
>Not critical, but might also be worth thinking about how this would interact with an interceptor (eg #4018), since that is on the table for 0.11, if we do go ahead with this approach.

@carlaKC I think the interceptor skips local payments.

File: htlcswitch/switch.go [hunk 4]
```diff
@@ -1263,7 +1251,8 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 		// A blank IncomingChanID in a circuit indicates that it is a pending
 		// user-initiated payment.
 		if packet.incomingChanID == hop.Source {
-			return s.handleLocalDispatch(packet)
+			s.wg.Add(1)
+			go s.handleLocalResponse(packet)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go:1188]: Good catch, fixed

File: htlcswitch/switch.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go:467]: This was pre-existing, I just moved the code. Will tack on another commit to change the behavior.

File: htlcswitch/interfaces.go [hunk 5]
```diff
@@ -68,7 +68,7 @@ type ChannelLink interface {
 	//
 	// NOTE: This function MUST be non-blocking (or block as little as
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/interfaces.go:69]: updated

File: htlcswitch/link.go (hunk 1 above)
Comment (joostjager) [reply htlcswitch/link.go:354]: fixed, good suggestion

File: htlcswitch/link.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/link.go:410]: the other downstream chan isn't initialized here

**Comment by joostjager** [comment]
Todo: itest fails because link does not cancel back local htlcs when commitment is full

File: htlcswitch/switch.go [hunk 6]
```diff
@@ -451,7 +451,26 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
+		log.Errorf("unable to commit circuit in switch: %v", err)
+		return err
+	}
+
+	// Drop duplicate packet if it has already been seen.
+	switch {
+	case len(actions.Drops) == 1:
+		return ErrDuplicateAdd
+
+	case len(actions.Fails) == 1:
```
Comment (Roasbeef) [inline htlcswitch/switch.go:470]: What about the logic to propagate the error backwards? I think as is, this'll change the error sent back to mission control, which may result in unexpected pruning behavior.

File: htlcswitch/switch.go [hunk 7]
```diff
@@ -470,7 +470,28 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
 	// Send packet to link.
 	packet.circuit = circuit
 
-	return s.route(packet)
+	// User have created the htlc update therefore we should find the
```
Comment (Roasbeef) [inline htlcswitch/switch.go:477]: The user has*

File: htlcswitch/switch.go [hunk 8]
```diff
@@ -470,7 +470,28 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
+	// appropriate channel link and send the payment over this link.
+	link, linkErr := s.handleLocalAddHTLC(packet, htlc)
```
Comment (Roasbeef) [inline htlcswitch/switch.go:479]: Pre-existing, but this method name really doesn't match the operation. It's more like `localEligibleLink` or something like that.

File: htlcswitch/interfaces.go [hunk 9]
```diff
@@ -67,8 +67,8 @@ type ChannelLink interface {
 	// initially.
 	//
 	// NOTE: This function MUST be non-blocking (or block as little as
-	// possible).
-	HandleSwitchPacket(*htlcPacket) error
+	// possible) if the sync bool parameter is false.
```
Comment (Roasbeef) [inline htlcswitch/interfaces.go:70]: Should be a new sentence. Also no body at all in this commit message.

File: htlcswitch/link.go [hunk 10]
```diff
@@ -1114,6 +1126,10 @@ func (l *channelLink) htlcManager() {
 		case pkt := <-l.downstream:
 			l.handleDownstreamPkt(pkt)
 
+		case msg := <-l.localDownstream:
+			l.handleDownstreamPkt(msg.pkt)
```
Comment (Roasbeef) [inline htlcswitch/link.go:1130]: Should add basic unit test that covers this new code path.

**Comment by Roasbeef** [comment]
Agreed with Eugene, I was expecting a larger overhaul to make this even possible in the first place. Nicely done!

File: htlcswitch/switch.go (hunk 6 above)
Comment (joostjager) [reply htlcswitch/switch.go:470]: Yes, that was still missing. I mentioned that in the comment above. I hadn't requested re-review yet, but thanks anyway for taking another look.

Back-path has been added.

File: htlcswitch/switch.go (hunk 7 above)
Comment (joostjager) [reply htlcswitch/switch.go:477]: fixed

File: htlcswitch/interfaces.go (hunk 9 above)
Comment (joostjager) [reply htlcswitch/interfaces.go:70]: Not applicable anymore

File: htlcswitch/switch.go (hunk 8 above)
Comment (joostjager) [reply htlcswitch/switch.go:479]: Renamed to `getLocalLink`

File: htlcswitch/link.go (hunk 10 above)
Comment (joostjager) [reply htlcswitch/link.go]: It is covered, for example in `TestChannelLinkSingleHopPayment`

File: htlcswitch/interfaces.go (hunk 9 above)
Comment (joostjager) [reply htlcswitch/interfaces.go]: Bodies added to this commit and others

**Comment by joostjager** [comment]
itest is now stuck on `sphinx_replay_persistence`

**Comment by joostjager** [comment]
Integration test fixed

File: htlcswitch/switch.go [hunk 11]
```diff
@@ -67,6 +67,10 @@ var (
 	// ErrUnreadableFailureMessage is returned when the failure message
 	// cannot be decrypted.
 	ErrUnreadableFailureMessage = errors.New("unreadable failure message")
+
+	// ErrLocalAddFailed signals that the ADD htlc for a local payment
+	// failed to be processed.
+	ErrLocalAddFailed = errors.New("local add HTLC failed")
```
Comment (carlaKC) [inline htlcswitch/switch.go:73]: Does this need to be exported?

File: htlcswitch/link.go [hunk 12]
```diff
@@ -1254,80 +1254,88 @@ func (l *channelLink) randomFeeUpdateTimeout() time.Duration {
...
-		// mailbox, and the HTLC being added to the commitment state.
-		if l.cfg.HodlMask.Active(hodl.AddOutgoing) {
-			l.log.Warnf(hodl.AddOutgoing.Warning())
-			l.mailBox.AckPacket(pkt.inKey())
-			return
-		}
+// handleDownstreamUpdateAdd processes an UpdateAddHTLC packet sent from the
+// downstream HTLC Switch.
+func (l *channelLink) handleDownstreamUpdateAdd(pkt *htlcPacket) {
+	htlc := pkt.htlc.(*lnwire.UpdateAddHTLC)
```
Comment (Roasbeef) [inline htlcswitch/link.go:1277]: Should check the second value here as a safety guard.

File: htlcswitch/switch_test.go [hunk 13]
```diff
@@ -2139,7 +2137,7 @@ func TestSwitchSendPayment(t *testing.T) {
 		},
 	}
 
-	if err := s.forward(packet); err != nil {
+	if err := <-s.ForwardPackets(nil, packet); err != nil {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch_test.go:2140]: Timeout added

File: htlcswitch/link.go [hunk 14]
```diff
@@ -2329,6 +2353,33 @@ func (l *channelLink) HandleSwitchPacket(pkt *htlcPacket) error {
...
 
+// HandleLocalAddPacket handles a locally-initiated UpdateAddHTLC packet. It
+// will be processed synchronously.
+//
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) HandleLocalAddPacket(pkt *htlcPacket) error {
+	l.log.Tracef("received switch packet outkey=%v", pkt.outKey())
+
+	// Create a buffered result channel to prevent the link from blocking.
+	errChan := make(chan error, 1)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/link.go:2364]: Yes, the former. If this function `HandleLocalAddPacket` quits, the link event loop can't deliver the message anymore.

File: htlcswitch/link.go (hunk 12 above)
Comment (joostjager) [reply htlcswitch/link.go:1277]: Added

File: htlcswitch/switch.go (hunk 11 above)
Comment (joostjager) [reply htlcswitch/switch.go:73]: Not strictly, but this error is returned to other sub-systems. By exporting it, those could check for it without string matching.

**Pull request merged**